### PR TITLE
[Core] Fix the segfault from Opencensus upon shutdown

### LINKED ISSUE #36885: [Core] OpenCensus segfault upon shutdown
### What happened + What you expected to happen

```
*** SIGSEGV received at time=1687888695 on cpu 0 ***
PC: @     0x7fd26334557f  (unknown)  opencensus::proto::metrics::v1::LabelKey::ByteSizeLong()
    @     0x7fd286b37420  974059328  (unknown)
    @     0x7fd2633507a0         48  opencensus::proto::metrics::v1::Metric::ByteSizeLong()
    @     0x7fd263306d0d         48  ray::rpc::ReportOCMetricsRequest::ByteSizeLong()
    @     0x7fd26313a5e8        208  grpc::GenericSerialize<>()
    @     0x7fd26313abe4        176  grpc::internal::ClientAsyncResponseReaderHelper::SetupRequest<>()
    @     0x7fd2632fbbcf        144  ray::rpc::ReporterService::Stub::PrepareAsyncReportOCMetricsRaw()
    @     0x7fd263206f2e         32  ray::rpc::ReporterService::Stub::PrepareAsyncReportOCMetrics()
    @     0x7fd263208790        208  ray::rpc::ClientCallManager::CreateCall<>()
    @     0x7fd263209e96        160  ray::rpc::MetricsAgentClient::ReportOCMetrics()
    @     0x7fd26320b1bb        384  ray::stats::OpenCensusProtoExporter::ExportViewData()
    @     0x7fd26372a96c        128  opencensus::stats::StatsExporterImpl::Export()
    @     0x7fd262fd1e21        432  ray::core::CoreWorker::Disconnect()
    @     0x7fd262ff1c31        128  ray::core::CoreWorkerProcessImpl::ShutdownDriver()
    @     0x7fd262ff1d9b         80  ray::core::CoreWorkerProcess::Shutdown()
    @     0x7fd262e3347f         32  __pyx_pw_3ray_7_raylet_10CoreWorker_3shutdown()
    @           0x4fcc35  (unknown)  method_vectorcall_NOARGS
    @ ... and at least 1 more frames
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361: *** SIGSEGV received at time=1687888695 on cpu 0 ***
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361: PC: @     0x7fd26334557f  (unknown)  opencensus::proto::metrics::v1::LabelKey::ByteSizeLong()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd286b37420  974059328  (unknown)
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd2633507a0         48  opencensus::proto::metrics::v1::Metric::ByteSizeLong()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd263306d0d         48  ray::rpc::ReportOCMetricsRequest::ByteSizeLong()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd26313a5e8        208  grpc::GenericSerialize<>()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd26313abe4        176  grpc::internal::ClientAsyncResponseReaderHelper::SetupRequest<>()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd2632fbbcf        144  ray::rpc::ReporterService::Stub::PrepareAsyncReportOCMetricsRaw()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd263206f2e         32  ray::rpc::ReporterService::Stub::PrepareAsyncReportOCMetrics()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd263208790        208  ray::rpc::ClientCallManager::CreateCall<>()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd263209e96        160  ray::rpc::MetricsAgentClient::ReportOCMetrics()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd26320b1bb        384  ray::stats::OpenCensusProtoExporter::ExportViewData()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd26372a96c        128  opencensus::stats::StatsExporterImpl::Export()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd262fd1e21        432  ray::core::CoreWorker::Disconnect()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd262ff1c31        128  ray::core::CoreWorkerProcessImpl::ShutdownDriver()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd262ff1d9b         80  ray::core::CoreWorkerProcess::Shutdown()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @     0x7fd262e3347f         32  __pyx_pw_3ray_7_raylet_10CoreWorker_3shutdown()
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @           0x4fcc35  (unknown)  method_vectorcall_NOARGS
[2023-06-27 10:58:15,870 E 1808 1808] logging.cc:361:     @ ... and at least 1 more frames
Fatal Python error: Segmentation fault
​
Stack (most recent call first):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/_private/worker.py", line 1712 in shutdown
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/_private/client_mode_hook.py", line 103 in wrapper
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/_private/worker.py", line 1101 in __exit__
  File "/tmp/ray/session_2023-06-27_10-37-10_364343_102/runtime_resources/working_dir_files/gs_lmi-anyscale_production_2023-06-27T17_13_17_00_00_querytag-workflow/querytag/job.py", line 387 in tagging_update_pass
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/decorators.py", line 26 in new_func
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/core.py", line 760 in invoke
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/core.py", line 1404 in invoke
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/core.py", line 1657 in invoke
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/core.py", line 1055 in main
  File "/home/ray/anaconda3/lib/python3.9/site-packages/click/core.py", line 1130 in __call__
  File "/tmp/ray/session_2023-06-27_10-37-10_364343_102/runtime_resources/working_dir_files/gs_lmi-anyscale_production_2023-06-27T17_13_17_00_00_querytag-workflow/querytag/job.py", line 393 in <module>
  File "/home/ray/anaconda3/lib/python3.9/runpy.py", line 87 in _run_code
  File "/home/ray/anaconda3/lib/python3.9/runpy.py", line 197 in _run_module_as_main
Segmentation fault (core dumped)
```

### Versions / Dependencies

2.5.1

### Reproduction script

n/a

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Seems like client_ in the metrics exporter is not lock-protected, but it is accessed by 2 thread (by a metrics export thread + task execution thread when ExportNow is called). I added a lock. I am not 100% sure if this will fix the issue because I was unable to reproduce the issue, but we can ask the user to try after merging this PR. 

## Related issue number

Fixes https://github.com/ray-project/ray/issues/36885

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker_process.cc [hunk 1]
```diff
@@ -146,7 +146,6 @@ CoreWorkerProcessImpl::CoreWorkerProcessImpl(const CoreWorkerOptions &options)
 
 CoreWorkerProcessImpl::~CoreWorkerProcessImpl() {
   RAY_LOG(INFO) << "Destructing CoreWorkerProcessImpl. pid: " << getpid();
-  RAY_LOG(DEBUG) << "Stats stop in core worker.";
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker_process.cc:149]: we added the equivalent logs to stats::Shutdown

**Review by rickyyx — APPROVED** [review]
Could we do benchmarks on it? 

And looks like from the stacktrace, the SIGSEV is on the request object? Might be an opencensus bug?

**Comment by rkooo567** [comment]
> And looks like from the stacktrace, the SIGSEV is on the request object? Might be an opencensus bug?

We are sending a request actually, and the problem is from protobuf, so it is unlikely related to Opencensus. It should be something shutdown ordering issue. I am not sure if my PR will actually fix the following issue, but client is not protected properly, and we should fix it anyway.

**Comment by rickyyx** [comment]
I am a bit concerned of the potential perf impact from the locks?

**Comment by rkooo567** [comment]
@rickyyx the report code happens in a separate thread, and the main thread is only locked upon exit, so it should be okay. Let me know if you have other parts that could be problematic...

**Comment by zhe-thoughts** [comment]
@rkooo567 is there a cherry-pick PR? Thanks

**Comment by rkooo567** [comment]
created here; https://github.com/ray-project/ray/pull/37311

**Pull request merged**